lib/spack/docs/packaging_guide.rst

```diff
@@ -1769,6 +1769,70 @@ Python's ``setup_dependent_environment`` method also sets up some
 other variables, creates a directory, and sets up the ``PYTHONPATH``
 so that dependent packages can find their dependencies at build time.
 
+.. _dependency-searchpaths:
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Custom ``include_paths`` and ``lib_paths``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Many packages install library and include files into subdirectories under
+Spack's installation prefix.  For example, the libelf package installs
+its include files into the ``include/libelf`` directory instead of the
+more common ``include``.  This can break dependent packages that make
+assumptions about how their dependency's files are laid out.
+
+Spack allows packages to describe installation subdirectories using the
+``include_paths`` and ``lib_paths`` properties.  Set ``include_paths`` to
```

Comment:
Frankly, this seems to implement exactly the same functionality of `LibraryList` and `HeaderList` classes, which are used in, for example,

```
@property
    def libs(self):
```

My preference would be that we use those more natively, because that's the interface in place already and it addresses exactly the issue

> break dependent packages that make
 assumptions about how their dependency's files are laid out.

So maybe, instead of adding another way to achieve the same goal, internally Spack environment and alike should more rely on `LibraryList` and `HeaderList`.

Reply:
those two classes can also be extended further, if there is a need.

Reply:
p.s. `LibraryList` and `HeaderList` were not as developed back at the end of 2016, when this PR was created.

Reply:
I still need to take a look at the code in details, but at a first glance I tend to agree with the considerations above. The functionality in the current state seems a duplicate of what `libs` and `headers` methods are doing.  Furthermore, I tried "as a user" to:
```console
$ spack install libdwarf ^libelf
```
and to my surprise it works on `develop` and not in this PR. The machines I used are:
```console
$ uname -a
Linux nuvolari 4.4.0-97-generic #120~14.04.1-Ubuntu SMP Wed Sep 20 15:53:13 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```
and
```console
$ uname -a
Linux fidis 3.10.0-514.26.2.el7.x86_64 #1 SMP Fri Jun 30 05:26:04 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```